cunumeric/module.py

```diff
@@ -2381,6 +2382,59 @@ def indices(
         return res_array
 
 
+def mask_indices(n: int, mask_func, k: int = 0) -> tuple[ndarray, ...]:
+    """
+    Return the indices to access (n, n) arrays, given a masking function.
+
+    Assume `mask_func` is a function that, for a square array a of size
+    ``(n, n)`` with a possible offset argument `k`, when called as
+    ``mask_func(a, k)`` returns a new array with zeros in certain locations
+    (functions like `triu` or `tril` do precisely this). Then this function
+    returns the indices where the non-zero values would be located.
+
+    Parameters
+    ----------
+    n : int
+        The returned indices will be valid to access arrays of shape (n, n).
+    mask_func : callable
+        A function whose call signature is similar to that of `triu`, `tril`.
+        That is, ``mask_func(x, k)`` returns a boolean array, shaped like `x`.
+        `k` is an optional argument to the function.
+    k : scalar
+        An optional argument which is passed through to `mask_func`. Functions
+        like `triu`, `tril` take a second argument that is interpreted as an
+        offset.
+
+    Returns
+    -------
+    indices : tuple of arrays.
+        The `n` arrays of indices corresponding to the locations where
+        ``mask_func(np.ones((n, n)), k)`` is True.
+
+    See Also
+    --------
+    numpy.mask_indices
+
+    Notes
+    -----
+    will work on GPU only when `mask_func` is a cuNumeric function, otherwise will
+    try to materialize GPU data on the host
+
+    Availability
+    --------
+    Multiple GPUs (when `mask_func` is a cunumeric function) , Multiple CPUs
+    """
+    a = ones((n, n), dtype=bool)
+    if not is_implemented(mask_func):
```

Comment:
This is more strict than it could be. if `mask_func` simply calls cuNumeric functions internally, it's safe to use even though it's a Python function.

Reply:
OK, removed the check

Reply:
I don't remember whether the removed check simply raised a warning or an exception (and I can't find it since you force pushed it), but if it was the former, let's put it back and print out the warning from the docstring.

Reply:
I realized I shouldn't do a force push a minute after I did that, sorry. It was a warning. I put it back and corrected dock-string

Reply:
Simply checking `is_cunumeric` can result in false positives, e.g. if the user passes a lambda, which is otherwise ok as it only calls cunumeric functions:

```
print(cn.mask_indices(3, lambda x,k: cn.tril(x,k)))
```

Given that this check is prone to false positives like that, I would rather avoid it. The user will get notified that their function is unimplemented when that function is invoked later anyway.

Reply:
@manolis, do you suggest I just remove the warning?

Reply:
That's my suggestion, but let's see what @magnatelee also has to say, since he originally advocated for this warning.

Reply:
I was worried about the case where the Python function passes a cuNumeric array to something that expects a NumPy array, which would end up materializing it on a single memory. This wouldn't get caught by our logic for unimplemented functions. The suggested check is sound and intended to make the users think hard on what the passed Python function would do. Alternatively, we could raise a warning in `DeferredArray.__numpy_array__`.

Reply:
OK, then let's put back the warning, but change the wording to reflect that the check is not precise:

```
    if not is_implemented(mask_func):
         runtime.warn(
             "Calling non-cuNumeric functions in mask_func can result in bad "
             "performance",
             category=UserWarning,
         )
```

I would prefer to have a general "numpy fallback" warning in `__numpy_array__`, but that would require a more thorough investigation, to make sure we don't emit extraneous warnings, so I'd rather we don't tackle that in this PR.